### ISSUE TITLE
Mixed type project (webpack and rust)

### ISSUE THREAD

**Issue by ksesong** [cloudflare/wrangler#232]
I have a mixed project using `webpack` and a rust package compiled beforehand with `wasm-bindgen`. Everything is handled by a bash script, and I wondered whether I could simplify this workflow by using `wrangler`.

I tried using the `rust` template, edited `wrangler.toml` to set `type = "webpack"` and tried importing a random package, but the preview yields `wasm_bindgen is not defined` as an error. Is this process supported? or is it too early for now?

**Comment by xtuc** [comment]
Have you tried using https://github.com/wasm-tool/wasm-pack-plugin? You should be able to import the wasm module via webpack.

**Comment by ksesong** [comment]
I've been trying a few things, for example, copying `https://github.com/xtuc/wrangler-js-demo/tree/master/rust` (which uses `wasm-pack-plugin`) and running `wrangler preview get` in the directory gives me this error `👷‍♀️ Your worker responded with: ReferenceError: wasmprogram is not defined`.
The created wasm module is named `module.wasm` in the `worker` directory, but the `metadata.json` references a different file:

```
                {
                    "body_part": "script",
                    "binding": {
                        "name": "wasmprogram",
                        "type": "wasm_module",
                        "part": "wasmprogram"
                    }
                }
```

Is there any barebone template that I can use?

**Comment by granjef3** [comment]
Right now a few things are broken in the latest released version of wrangler (namely, the binding upload format) for this to work. These issues should be fixed soon! However, I will give instructions to get this working today, as well as instructions to go back to the next stable release of wrangler when it comes out.

- `git clone` the master branch of wrangler.
- run `cargo uninstall wrangler` to get rid of the current released version of wrangler.
- run `cargo install --path .` from inside the wrangler directory you cloned, this will install the master version of wrangler that fixes the binding format
   however, wrangler still pulls in some dependencies when installed this way, and we need to update one of them manually (wranglerjs, the glue layer used to call webpack) as the released one is incompatible with the master branch
- go to the [cache directory](https://docs.rs/dirs/2.0.1/dirs/fn.cache_dir.html) on your machine, where a `.wrangler` directory will be located.
- inside `.wrangler` replace `index.js` with the latest version from the current master version of wrangler using `rm index.js && wget https://raw.githubusercontent.com/cloudflare/wrangler/master/wranglerjs/index.js`

Now you're on the latest wrangler. In your project you need to replace the contents of `worker/metadata.json` with 
```
{
    "body_part": "script",
    "bindings": [
        {
            "name": "wasmprogram",
            "type": "wasm_module",
            "part": "wasmprogram"
        }
    ]
}
```

add to your `webpack.config.js` the following settings. You'll need to add `wasm-pack-plugin` to your package.json as a devDependency.

```
const WasmPackPlugin = require("@wasm-tool/wasm-pack-plugin");

module.exports = {
  entry: {
    index: <wherever the entry point javascript file for your worker is>
  },
  plugins: [
    new WasmPackPlugin({
      crateDirectory: __dirname,
      extraArgs: "--out-name index"
    }),
  ]
};
```

Finally, in the JS portion of your app, use `let wasm = await import('<relative path to the pkg directory from that file>'`. wasm will be an object that contains the functions you exported using `wasm-bindgen`, you can look inside `pkg/index.js` to see how it's created after running `wrangler build`

When a new version of wrangler is released that fixes rust+webpack projects, you'll have to run `cargo uninstall wrangler` and delete the `.wrangler` directory in the cache dir (only to go back to stable, NOT for normal upgrades).

**Comment by ashleymichal** [comment]
Some of this work should be made easier by the work we are doing to fix bindings in general.

**Comment by ashleymichal** [comment]
removing this from the Refactor Bindings, Bundling milestone, as what we really need is a webpack config template for this.

**Comment by dlackty** [comment]
Though @xortive provided a workaround at this moment, I spend few hours and still couldn't get it working.

From the README, is describes:
> webpack: This project contains any number of JavaScript files or Rust/C/C++ files that compile to WebAssembly. Rust files will be built with wasm-pack. This project type uses webpack and webpack plugins in the background to build your worker.

Just wanna make sure that `wasm-pack-plugin` is still required?

**Comment by xortive** [comment]
@dlackty yes, `wasm-pack-plugin` is still required to call `wasm-pack` when using the webpack project. The README is inaccurate here, I'll update this issue to reflect that

**Comment by ashleygwilliams** [comment]
i think we keep missing this issue because it's filed here and not on the docs repo. now that that repo is open source, i'm going to move this to that repo!
